File: src/main/java/org/modelingvalue/nelumbo/InferResult.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2025 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo;

import org.modelingvalue.collections.Collection;
import org.modelingvalue.collections.List;
import org.modelingvalue.collections.Map;
import org.modelingvalue.collections.Set;

public interface InferResult {

    Set<Predicate> facts();

    Set<Predicate> falsehoods();

    default Collection<Predicate> allFacts() {
        return facts();
    }

    default Collection<Predicate> allFalsehoods() {
        return falsehoods();
    }

    boolean completeFacts();

    boolean completeFalsehoods();

    default Predicate predicate() {
        return null;
    }

    default Set<Predicate> cycles() {
        return Set.of();
    }

    default List<Predicate> stackOverflow() {
        return null;
    }

    default boolean hasCycleWith(Predicate predicate) {
        return cycles().contains(predicate);
    }

    default boolean hasStackOverflow() {
        return stackOverflow() != null;
    }

    default boolean isTrueCC() {
        return allFalsehoods().isEmpty() && !allFacts().isEmpty() && completeFalsehoods() && completeFacts();
    }

    default boolean isFalseCC() {
        return allFacts().isEmpty() && !allFalsehoods().isEmpty() && completeFacts() && completeFalsehoods();
    }

    default boolean isComplete() {
        return completeFacts() || completeFalsehoods();
    }

    default boolean isEmpty() {
        return allFacts().isEmpty() && allFalsehoods().isEmpty() && cycles().isEmpty();
    }

    default Set<Map<Variable, Object>> trueBindings() {
        return allFacts().map(p -> p.getBinding(predicate()).removeAll(e -> e.getValue() instanceof Variable || e.getValue() instanceof Type)).asSet();
    }

    default Set<Map<Variable, Object>> falseBindings() {
        return allFalsehoods().map(p -> p.getBinding(predicate()).removeAll(e -> e.getValue() instanceof Variable || e.getValue() instanceof Type)).asSet();
    }

    static InferResult of(Set<Predicate> facts, boolean completeFacts, Set<Predicate> falsehoods, boolean completeFalsehoods, Set<Predicate> cycles) {
        return new InferResultImpl() {
            @Override
            public Set<Predicate> facts() {
                return facts;
            }

            @Override
            public Set<Predicate> falsehoods() {
                return falsehoods;
            }

            @Override
            public boolean completeFacts() {
                return completeFacts;
            }

            @Override
            public boolean completeFalsehoods() {
                return completeFalsehoods;
            }

            @Override
            public Set<Predicate> cycles() {
                return cycles;
            }
        };
    }

    static InferResult of(Predicate predicate, Set<Predicate> facts, boolean completeFacts, Set<Predicate> falsehoods, boolean completeFalsehoods, Set<Predicate> cycles) {
        return new InferResultImpl() {

            @Override
            public Predicate predicate() {
                return predicate;
            }

            @Override
            public Set<Predicate> facts() {
                return facts;
            }

            @Override
            public Set<Predicate> falsehoods() {
                return falsehoods;
            }

            @Override
            public boolean completeFacts() {
                return completeFacts;
            }

            @Override
            public boolean completeFalsehoods() {
                return completeFalsehoods;
            }

            @Override
            public Set<Predicate> cycles() {
                return cycles;
            }
        };
    }

    static InferResult of(Collection<Predicate> facts, boolean completeFacts, Collection<Predicate> falsehoods, boolean completeFalsehoods, Set<Predicate> cycles) {
        return new InferResultImpl() {
            @Override
            public Set<Predicate> facts() {
                return null;
            }

            @Override
            public Set<Predicate> falsehoods() {
                return null;
            }

            @Override
            public Collection<Predicate> allFacts() {
                return facts;
            }

            @Override
            public Collection<Predicate> allFalsehoods() {
                return falsehoods;
            }

            @Override
            public boolean completeFacts() {
                return completeFacts;
            }

            @Override
            public boolean completeFalsehoods() {
                return completeFalsehoods;
            }

            @Override
            public Set<Predicate> cycles() {
                return cycles;
            }

        };
    }

    static InferResult unknown(Predicate unknown) {
        return new InferResultImpl() {
            @Override
            public Set<Predicate> facts() {
                return Set.of();
            }

            @Override
            public Set<Predicate> falsehoods() {
                return Set.of();
            }

            @Override
            public boolean completeFacts() {
                return false;
            }

            @Override
            public boolean completeFalsehoods() {
                return false;
            }

            @Override
            public Predicate predicate() {
                return unknown;
            }
        };
    }

    static InferResult factsCI(Set<Predicate> facts) {
        return new InferResultImpl() {
            @Override
            public Set<Predicate> facts() {
                return facts;
            }

            @Override
            public Set<Predicate> falsehoods() {
                return Set.of();
            }

            @Override
            public boolean completeFacts() {
                return true;
            }

            @Override
            public boolean completeFalsehoods() {
                return false;
            }
        };
    }

    static InferResult factsIC(Set<Predicate> facts) {
        return new InferResultImpl() {
            @Override
            public Set<Predicate> facts() {
                return facts;
            }

            @Override
            public Set<Predicate> falsehoods() {
                return Set.of();
            }

            @Override
            public boolean completeFacts() {
                return false;
            }

            @Override
            public boolean completeFalsehoods() {
                return true;
            }
        };
    }

    static InferResult factsCC(Set<Predicate> facts) {
        return new InferResultImpl() {
            @Override
            public Set<Predicate> facts() {
                return facts;
            }

            @Override
            public Set<Predicate> falsehoods() {
                return Set.of();
            }

            @Override
            public boolean completeFacts() {
                return true;
            }

            @Override
            public boolean completeFalsehoods() {
                return true;
            }
        };
    }

    static InferResult falsehoodsCC(Set<Predicate> falsehoods) {
        return new InferResultImpl() {
            @Override
            public Set<Predicate> facts() {
                return Set.of();
            }

            @Override
            public Set<Predicate> falsehoods() {
                return falsehoods;
            }

            @Override
            public boolean completeFacts() {
                return true;
            }

            @Override
            public boolean completeFalsehoods() {
                return true;
            }
        };
    }

    static InferResult falsehoodsIC(Set<Predicate> falsehoods) {
        return new InferResultImpl() {
            @Override
            public Set<Predicate> facts() {
                return Set.of();
            }

            @Override
            public Set<Predicate> falsehoods() {
                return falsehoods;
            }

            @Override
            public boolean completeFacts() {
                return false;
            }

            @Override
            public boolean completeFalsehoods() {
                return true;
            }

        };
    }

    static InferResult falsehoodsCI(Predicate falsehood) {
        return new InferResultImpl() {
            @Override
            public Set<Predicate> facts() {
                return Set.of();
            }

            @Override
            public Set<Predicate> falsehoods() {
                return Set.of();
            }

            @Override
            public boolean completeFacts() {
                return true;
            }

            @Override
            public boolean completeFalsehoods() {
                return false;
            }

            @Override
            public Predicate predicate() {
                return falsehood;
            }

        };
    }

    static InferResult cycle(Set<Predicate> facts, Set<Predicate> falsehoods, Predicate predicate) {
        return new InferResultImpl() {
            @Override
            public Set<Predicate> facts() {
                return facts;
            }

            @Override
            public Set<Predicate> falsehoods() {
                return falsehoods;
            }

            @Override
            public boolean completeFacts() {
                return false;
            }

            @Override
            public boolean completeFalsehoods() {
                return false;
            }

            @Override
            public Set<Predicate> cycles() {
                return predicate.singleton();
            }
        };
    }

    static InferResult overflow(List<Predicate> overflow) {
        return new InferResultImpl() {
            @Override
            public Set<Predicate> facts() {
                return null;
            }

            @Override
            public Set<Predicate> falsehoods() {
                return null;
            }

            @Override
            public boolean completeFacts() {
                return false;
            }

            @Override
            public boolean completeFalsehoods() {
                return false;
            }

            @Override
            public List<Predicate> stackOverflow() {
                return overflow;
            }
        };
    }

    default InferResult addAnd(InferResult other) {
        List<Predicate> facts = Collection.concat(allFacts(), other.allFacts()).asList();
        boolean completeFacts = completeFacts() || other.completeFacts();
        List<Predicate> falsehoods = Collection.concat(allFalsehoods(), other.allFalsehoods()).asList();
        boolean completeFalsehoods = completeFalsehoods() && other.completeFalsehoods();
        Set<Predicate> cycles = cycles().addAll(other.cycles());
        return of(facts, completeFacts, falsehoods, completeFalsehoods, cycles);
    }

    default InferResult addOr(InferResult other) {
        List<Predicate> facts = Collection.concat(allFacts(), other.allFacts()).asList();
        boolean completeFacts = completeFacts() && other.completeFacts();
        List<Predicate> falsehoods = Collection.concat(allFalsehoods(), other.allFalsehoods()).asList();
        boolean completeFalsehoods = completeFalsehoods() || other.completeFalsehoods();
        Set<Predicate> cycles = cycles().addAll(other.cycles());
        return of(facts, completeFacts, falsehoods, completeFalsehoods, cycles);
    }

    default InferResult flipComplete() {
        return of(allFacts(), completeFalsehoods(), allFalsehoods(), completeFacts(), cycles());
    }

    default InferResult biimply(InferResult ruleResult, Rule rule) {
        if (checkConsistency(ruleResult) || ruleResult.checkConsistency(this)) {
            throw new InconsistencyException(rule, ruleResult, this);
        }
        Set<Predicate> facts = facts().addAll(ruleResult.facts());
        Set<Predicate> falsehoods = falsehoods().addAll(ruleResult.falsehoods());
        boolean completeFacts = completeFacts() || ruleResult.completeFacts();
        boolean completeFalsehoods = completeFalsehoods() || ruleResult.completeFalsehoods();
        Set<Predicate> cycles = cycles().addAll(ruleResult.cycles());
        return of(facts, completeFacts, falsehoods, completeFalsehoods, cycles);
    }

    default boolean checkConsistency(InferResult other) {
        return (other.completeFacts() && !facts().allMatch(other.facts()::contains)) || //
                (other.completeFalsehoods() && !falsehoods().allMatch(other.falsehoods()::contains));
    }

    default InferResult cast(Predicate to) {
        return of(cast(facts(), to), completeFacts(), cast(falsehoods(), to), completeFalsehoods(), cycles());
    }

    default InferResult predicate(Predicate predicate) {
        return of(predicate, facts(), completeFacts(), falsehoods(), completeFalsehoods(), cycles());
    }

    static Set<Predicate> cast(Set<Predicate> set, Predicate to) {
        return set.replaceAll(p -> p.equals(to) ? to : to.castFrom(p));
    }

    abstract class InferResultImpl implements InferResult {
        @Override
        public String toString() {
            List<Predicate> overflow = stackOverflow();
            if (overflow != null) {
                return overflow.toString().substring(4);
            } else {
                String cycleString = "";
                if (!cycles().isEmpty()) {
                    cycleString = cycles().toString().substring(3);
                    cycleString = "{" + cycleString.substring(1, cycleString.length() - 1) + "}";
                }
                return toString(trueBindings(), completeFacts()) + toString(falseBindings(), completeFalsehoods()) + cycleString;
            }
        }

        private String toString(Set<Map<Variable, Object>> binding, boolean complete) {
            List<String> stringList = binding.map(m -> {
                String map = m.toString();
                return "(" + map.substring(4, map.length() - 1) + ")";
            }).sorted().asList();
            String result = stringList.toString().substring(4);
            return complete ? result : result.substring(0, result.length() - 1) + (binding.isEmpty() ? "..]" : ",..]");
        }

        @Override
        public int hashCode() {
            int h = (completeFacts() ? 3 : 0) + (completeFalsehoods() ? 7 : 0);
            return h + allFacts().hashCode() ^ allFalsehoods().hashCode();
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) {
                return true;
            } else if (obj == null) {
                return false;
            } else if (!(obj instanceof InferResult other)) {
                return false;
            } else {
                return allFacts().equals(other.allFacts()) && completeFacts() == other.completeFacts() && //
                        allFalsehoods().equals(other.allFalsehoods()) && completeFalsehoods() == other.completeFalsehoods() && //
                        cycles().equals(other.cycles());
            }
        }
    }

}

File: web-ts/src/InferResult.ts
/**
 * InferResult - three-state logic result.
 * @JAVA_REF org.modelingvalue.nelumbo.InferResult
 */

import { List, Map, Set } from 'immutable';
import { Variable } from './Variable';
import { Type } from './Type';
import type { Predicate } from './logic/Predicate';

/**
 * InferResult - represents the result of logical inference.
 * @JAVA_REF org.modelingvalue.nelumbo.InferResult
 */
export class InferResult {
  private readonly _facts: Set<Predicate>;
  private readonly _falsehoods: Set<Predicate>;
  private readonly _allFacts: List<Predicate> | null;
  private readonly _allFalsehoods: List<Predicate> | null;
  private readonly _cycles: Set<Predicate>;
  private readonly _completeFacts: boolean;
  private readonly _completeFalsehoods: boolean;
  private readonly _predicate: Predicate | null;
  private readonly _stackOverflow: List<Predicate> | null;
  private readonly _unresolvable: boolean;

  private constructor(
    facts: Set<Predicate>,
    falsehoods: Set<Predicate>,
    cycles: Set<Predicate>,
    completeFacts: boolean,
    completeFalsehoods: boolean,
    predicate: Predicate | null = null,
    stackOverflow: List<Predicate> | null = null,
    unresolvable: boolean = false,
    allFacts: List<Predicate> | null = null,
    allFalsehoods: List<Predicate> | null = null
  ) {
    this._facts = facts;
    this._falsehoods = falsehoods;
    this._cycles = cycles;
    this._completeFacts = completeFacts;
    this._completeFalsehoods = completeFalsehoods;
    this._predicate = predicate;
    this._stackOverflow = stackOverflow;
    this._unresolvable = unresolvable;
    this._allFacts = allFacts;
    this._allFalsehoods = allFalsehoods;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#of(Set, boolean, Set, boolean, Set)
  static of(
    facts: Set<Predicate>,
    completeFacts: boolean,
    falsehoods: Set<Predicate>,
    completeFalsehoods: boolean,
    cycles: Set<Predicate>
  ): InferResult {
    return new InferResult(facts, falsehoods, cycles, completeFacts, completeFalsehoods);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#of(Predicate, Set, boolean, Set, boolean, Set)
  static ofWithPredicate(
    predicate: Predicate,
    facts: Set<Predicate>,
    completeFacts: boolean,
    falsehoods: Set<Predicate>,
    completeFalsehoods: boolean,
    cycles: Set<Predicate>
  ): InferResult {
    return new InferResult(facts, falsehoods, cycles, completeFacts, completeFalsehoods, predicate);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#of(Collection, boolean, Collection, boolean, Set)
  static ofFromCollections(
    facts: List<Predicate>,
    completeFacts: boolean,
    falsehoods: List<Predicate>,
    completeFalsehoods: boolean,
    cycles: Set<Predicate>
  ): InferResult {
    return new InferResult(
      Set<Predicate>(), Set<Predicate>(), cycles,
      completeFacts, completeFalsehoods,
      null, null, false,
      facts, falsehoods
    );
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#factsCI(Set)
  static factsCI(facts: Set<Predicate>): InferResult {
    return new InferResult(facts, Set(), Set(), true, false);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#factsIC(Set)
  static factsIC(facts: Set<Predicate>): InferResult {
    return new InferResult(facts, Set(), Set(), false, true);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#factsCC(Set)
  static factsCC(facts: Set<Predicate>): InferResult {
    return new InferResult(facts, Set(), Set(), true, true);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#falsehoodsCC(Set)
  static falsehoodsCC(falsehoods: Set<Predicate>): InferResult {
    return new InferResult(Set(), falsehoods, Set(), true, true);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#falsehoodsIC(Set)
  static falsehoodsIC(falsehoods: Set<Predicate>): InferResult {
    return new InferResult(Set(), falsehoods, Set(), false, true);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#falsehoodCI(Predicate)
  static falsehoodCI(falsehood: Predicate): InferResult {
    return new InferResult(Set(), Set(), Set(), true, false, falsehood);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#falsehoodsII(Set)
  static falsehoodsII(falsehoods: Set<Predicate>): InferResult {
    return new InferResult(Set(), falsehoods, Set(), false, false);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#unknown(Predicate)
  static unknown(predicate: Predicate): InferResult {
    return new InferResult(Set(), Set(), Set(), false, false, predicate);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#unresolvable(Predicate)
  static unresolvable(predicate: Predicate): InferResult {
    return new InferResult(Set(), Set(), Set(), false, false, predicate, null, true);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#cycle(Set, Set, Predicate)
  static cycle(facts: Set<Predicate>, falsehoods: Set<Predicate>, predicate: Predicate): InferResult {
    return new InferResult(facts, falsehoods, predicate.singleton(), false, false);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#overflow(List)
  static overflow(overflow: List<Predicate>): InferResult {
    return new InferResult(Set(), Set(), Set(), false, false, null, overflow);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult.UNRESOLVABLE
  static readonly UNRESOLVABLE = new InferResult(Set(), Set(), Set(), true, true, null, null, true);

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#facts()
  facts(): Set<Predicate> {
    return this._facts;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#falsehoods()
  falsehoods(): Set<Predicate> {
    return this._falsehoods;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#allFacts()
  allFacts(): Iterable<Predicate> & { isEmpty(): boolean; equals(other: unknown): boolean; map<T>(fn: (v: Predicate) => T): { toArray(): T[] } } {
    if (this._allFacts !== null) {
      return this._allFacts as any;
    }
    return this._facts as any;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#allFalsehoods()
  allFalsehoods(): Iterable<Predicate> & { isEmpty(): boolean; equals(other: unknown): boolean; map<T>(fn: (v: Predicate) => T): { toArray(): T[] } } {
    if (this._allFalsehoods !== null) {
      return this._allFalsehoods as any;
    }
    return this._falsehoods as any;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#cycles()
  cycles(): Set<Predicate> {
    return this._cycles;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#completeFacts()
  completeFacts(): boolean {
    return this._completeFacts;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#completeFalsehoods()
  completeFalsehoods(): boolean {
    return this._completeFalsehoods;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#isComplete()
  isComplete(): boolean {
    return this._completeFacts || this._completeFalsehoods;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#predicate()
  predicateOf(): Predicate | null {
    return this._predicate;
  }

  /**
   * Get or set the predicate.
   * @JAVA_REF org.modelingvalue.nelumbo.InferResult#predicate() and InferResult#predicate(Predicate)
   */
  predicate(pred?: Predicate): InferResult | Predicate | null {
    if (pred !== undefined) {
      return InferResult.ofWithPredicate(pred, this._facts, this._completeFacts, this._falsehoods, this._completeFalsehoods, this._cycles);
    }
    return this._predicate;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#unresolvable()
  isUnresolvable(): boolean {
    return this._unresolvable;
  }

  /**
   * Java-compatible unresolvable() default method.
   * @JAVA_REF org.modelingvalue.nelumbo.InferResult#unresolvable()
   */
  unresolvable(): boolean {
    return this._unresolvable;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#hasCycleWith(Predicate)
  hasCycleWith(predicate: Predicate): boolean {
    return this._cycles.contains(predicate);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#stackOverflow()
  stackOverflow(): List<Predicate> | null {
    return this._stackOverflow;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#hasStackOverflow()
  hasStackOverflow(): boolean {
    return this._stackOverflow !== null;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#isTrueCC()
  isTrueCC(): boolean {
    return this.allFalsehoods().isEmpty() && !this.allFacts().isEmpty() && this._completeFalsehoods && this._completeFacts;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#isFalseCC()
  isFalseCC(): boolean {
    return this.allFacts().isEmpty() && !this.allFalsehoods().isEmpty() && this._completeFacts && this._completeFalsehoods;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#trueBindings()
  trueBindings(): Set<Map<Variable, unknown>> {
    const pred = this._predicate;
    if (pred === null) return Set();
    const allF = this.allFacts();
    let result = Set<Map<Variable, unknown>>();
    for (const p of allF) {
      const binding = p.getBinding(pred);
      if (binding === null) {
        result = result.add(Map<Variable, unknown>());
      } else {
        const filtered = binding.filter((val) => !(val instanceof Variable) && !(val instanceof Type));
        result = result.add(filtered);
      }
    }
    return result;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#falseBindings()
  falseBindings(): Set<Map<Variable, unknown>> {
    const pred = this._predicate;
    if (pred === null) return Set();
    const allFH = this.allFalsehoods();
    let result = Set<Map<Variable, unknown>>();
    for (const p of allFH) {
      const binding = p.getBinding(pred);
      if (binding === null) {
        result = result.add(Map<Variable, unknown>());
      } else {
        const filtered = binding.filter((val) => !(val instanceof Variable) && !(val instanceof Type));
        result = result.add(filtered);
      }
    }
    return result;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#cast(Predicate)
  cast(to: Predicate): InferResult {
    const castFacts = this._facts.map(p => p.equals(to) ? to : to.castFrom(p)).toSet();
    const castFalsehoods = this._falsehoods.map(p => p.equals(to) ? to : to.castFrom(p)).toSet();
    return InferResult.of(castFacts, this._completeFacts, castFalsehoods, this._completeFalsehoods, this._cycles);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#add(InferResult)
  add(other: InferResult): InferResult {
    const facts = List<Predicate>([...this.allFacts(), ...other.allFacts()]);
    const falsehoods = List<Predicate>([...this.allFalsehoods(), ...other.allFalsehoods()]);
    const completeFacts = this._completeFacts && other._completeFacts;
    const completeFalsehoods = this._completeFalsehoods && other._completeFalsehoods;
    const cycles = this._cycles.union(other._cycles);
    return InferResult.ofFromCollections(facts, completeFacts, falsehoods, completeFalsehoods, cycles);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#flipComplete()
  flipComplete(): InferResult {
    const allF = this.allFacts();
    const allFH = this.allFalsehoods();
    if (this._allFacts !== null || this._allFalsehoods !== null) {
      return InferResult.ofFromCollections(
        List<Predicate>([...allF]),
        this._completeFalsehoods,
        List<Predicate>([...allFH]),
        this._completeFacts,
        this._cycles
      );
    }
    return InferResult.of(this._facts, this._completeFalsehoods, this._falsehoods, this._completeFacts, this._cycles);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult#complete()
  complete(): InferResult {
    if (this._allFacts !== null || this._allFalsehoods !== null) {
      return InferResult.ofFromCollections(
        List<Predicate>([...this.allFacts()]),
        true,
        List<Predicate>([...this.allFalsehoods()]),
        true,
        this._cycles
      );
    }
    return InferResult.of(this._facts, true, this._falsehoods, true, this._cycles);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult.InferResultImpl#equals(Object)
  equals(other: InferResult): boolean {
    if (this === other) return true;
    if (other === null || other === undefined) return false;
    const thisAllFacts = this.allFacts();
    const otherAllFacts = other.allFacts();
    const thisAllFalsehoods = this.allFalsehoods();
    const otherAllFalsehoods = other.allFalsehoods();
    return thisAllFacts.equals(otherAllFacts) &&
           this._completeFacts === other._completeFacts &&
           thisAllFalsehoods.equals(otherAllFalsehoods) &&
           this._completeFalsehoods === other._completeFalsehoods &&
           this._cycles.equals(other._cycles);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.InferResult.InferResultImpl#toString()
  toString(): string {
    if (this._stackOverflow !== null) {
      return this._stackOverflow.toArray().toString();
    }
    let cycleString = '';
    if (!this._cycles.isEmpty()) {
      cycleString = '{' + this._cycles.toArray().join(',') + '}';
    }
    const trueStr = this.bindingsToString(this.trueBindings(), this._completeFacts);
    const falseStr = this.bindingsToString(this.falseBindings(), this._completeFalsehoods);
    return trueStr + falseStr + cycleString;
  }

  private bindingsToString(bindings: Set<Map<Variable, unknown>>, complete: boolean): string {
    const strs = bindings.map(m => {
      const entries: string[] = [];
      m.forEach((v, k) => {
        entries.push(k + '=' + v);
      });
      return '(' + entries.join(',') + ')';
    }).toArray().sort();
    const result = '[' + strs.join(',') + ']';
    return complete ? result : result.slice(0, -1) + (bindings.isEmpty() ? '..]' : ',..]');
  }
}
